### ISSUE TITLE
Storage is configured but metrics are empty when mem_buf_limit reached

### ISSUE THREAD

**Issue by dmarkwat** [fluent/fluent-bit#1217]
## Bug Report

**Describe the bug**
I currently have fluentbit configured with both a forward and tcp input and 850M `mem_buf_limit` for each. Storage is configured, and both inputs are configured to use the filesystem. Due to an output plugin configured with unlimited retries (`Retry_Limit False`) causing errors and forcing fluentbit into a retry loop, the mem_buf_limit is getting reached on both input plugins.

But, I see no log messages related to the storage backlog being used, and the input records metrics all level-off (prometheus' `delta` function on the metric reads 0). I'd expect that when mem_buf_limit is hit, records would continue to flow but spill to disk. I see disk is being used, but on a 1 TB mount only `1782836` bytes are being consumed, per `df`, and both input plugins have stopped accepting more records, per the exported metrics. I've taken this to mean the input plugins have `paused`, but I would expect the filesystem buffer would prevent that.

Strangely, despite storage backlog being configured, I see no log messages related to its use except the initial "initializing" and configured size (`4.8M`) in journald.

Should records continue to spill to disk when the memory limits have been reached? Or is this not how fluentbit is designed to work?

**To Reproduce**
- Use the provided config
- Run ES on localhost but force it to reject messages by, e.g. making it a coordinator with no data nodes or stopping associated data nodes to force a "no shards" type error; in effect, forcing fluentbit to retry instead of "fail" the output
- Wait for the flow of records to reach mem_buf_limit and watch the input_records_total metric on the capped input plugin fall to zero despite available disk space

**Expected behavior**
When mem_buf_limit on an input plugin is reached, records would continue to flow but spill to disk and not count against the mem_buf_limit.

**Your Environment**
* Version used: `Fluent Bit v1.0.4`
* Configuration:
```
[SERVICE]
    Buffer_Path                 /var/data/buffer
    Buffer_Workers              10
    Config_Watch                off
    Daemon                      off
    Flush                       5
    Grace                       5
    HTTP_Listen                 0.0.0.0
    HTTP_Port                   8888
    HTTP_Server                 True
    Log_Level                   info
    Parsers_File                parsers.conf
    storage.path                /var/data/storage
    storage.sync                normal
    storage.checksum            off
    storage.backlog.mem_limit   5M
[INPUT]
    Mem_Buf_Limit   850M
    storage.type    filesystem
    Name              forward
    Listen            0.0.0.0
    Port              24224
    Buffer_Max_Size   512KB
    Buffer_Chunk_Size 32KB
[INPUT]
    Tag             logstash
    Mem_Buf_Limit   850M
    storage.type    filesystem
    Name        tcp
    Listen      0.0.0.0
    Port        5170
    Chunk_Size  32
    Buffer_Size 64

[FILTER]
    Match           *
    
    Name    record_modifier
    Remove_key @timestamp
[FILTER]
    Match           *
    
    Name    grep
    Exclude type ^syslog$

[OUTPUT]
    Match           *
    Retry_Limit     False
    Name                es
    Host                127.0.0.1
    Port                9201
    Buffer_Size         128KB
    Generate_ID         on
    Logstash_Format     on
    Include_Tag_Key     on
    tls             off
```
* Environment name and version (e.g. Kubernetes? What version?): n/a
* Server type and version: x86
* Operating System and version: RedHat 7.5
* Filters and plugins:
  - input: tcp, forward
  - filters: record_modifier, grep
  - output: es

**Additional context**
Records stop flowing to upstream agents with forward input plugins bc they stop accepting records despite available resources (in this case, disk). Yes, ES should be made to stop failing but in times when it can't come back quickly enough, we want to rely on fluentbit to buffer according to the specs we've configured the machine with.

**Comment by edsiper** [comment]
@dmarkwat Indeed that's a missing feature at the moment but the expectation when filesystem storage is defined. I will work on a solution.

**Comment by edsiper** [comment]
@dmarkwat I've pushed some improvements on the storage layer, if you can build from GIT master and test would be appreciated.

**Comment by dmarkwat** [comment]
@edsiper appreciate the attention to this! I ran a simple test today using the same tcp input and a bad http output to simulate the retry loop that caused the original problem. The storage layer appears to be working and the files keep growing in the `storage.path`.

I turned debug logging on to get a better sense of what it's doing and I have 3 observations:
1. in `top`, the reported `RES` memory is growing beyond the `Mem_Buf_Limit` of `10MB` I set on the output, growing up to 24MB before I killed the test; I'm planning to run this for longer to see how it behaves
2. the retry tasks appear to stack, possibly causing the memory to balloon; and I noticed that the ingress rate dropped the longer I ran the test
3. upon restart, my user default `ulimit -n` (1024) isn't enough file descriptors for 33M of data (not an issue with systemd `LimitNOFile=65536`, but just offering the data point)

As an aside, I couldn't get the metrics to report--not sure if something in those settings are broken or if I missed something. `top` and watching the dir size was enough to see that things were working, just with the above observations.

Here's the config I used:
```
[SERVICE]
    Flush 5
    Daemon off
    HTTP_Server  On
    HTTP_Listen  0.0.0.0
    HTTP_PORT    8888
    storage.path ./fluentbit-storage
    storage.backlog.mem_limit 5M
    Log_Level debug
[INPUT]
    name tcp
    storage.type filesystem
    Listen 127.0.0.1
    Port 5170
    Buffer_size 128
    Chunk_Size 32KB
    Mem_Buf_Limit 10MB
[OUTPUT]
    name http
    match *
    Retry_Limit False
    Host 127.0.0.1
    Port 80
```
